spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/publisher/PubSubPublisherTemplate.java

```diff
@@ -0,0 +1,130 @@
+ * Default implementation of {@link PubSubPublisherOperations}.
+ *
+ * <p>The main Google Cloud Pub/Sub integration component for publishing to topics.
+ *
+ * @author Vinicius Carvalho
+ * @author João André Martins
+ * @author Mike Eltsufin
+ * @author Chengyuan Zhao
+ * @author Doug Hoard
+ * @since 1.1.0
+ */
+public class PubSubPublisherTemplate implements PubSubPublisherOperations, InitializingBean {
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubPublisherTemplate.class);
+
+	private PubSubMessageConverter messageConverter = new SimplePubSubMessageConverter();
+
+	private final PublisherFactory publisherFactory;
+
+	/**
+	 * Default {@link PubSubPublisherTemplate} constructor that uses {@link SimplePubSubMessageConverter}
+	 * to serialize and deserialize payloads.
+	 *
+	 * @param publisherFactory the {@link com.google.cloud.pubsub.v1.Publisher} factory to
+	 *                         publish to topics.
+	 */
+	public PubSubPublisherTemplate(PublisherFactory publisherFactory) {
+		Assert.notNull(publisherFactory, "A valid PublisherFactory is required.");
+
+		this.publisherFactory = publisherFactory;
+	}
+
+	public PubSubMessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	public PubSubPublisherTemplate setMessageConverter(PubSubMessageConverter messageConverter) {
+		Assert.notNull(messageConverter, "A valid Pub/Sub message converter is required.");
+		this.messageConverter = messageConverter;
+
+		return this;
+	}
+
+	/**
+	 * Uses the configured message converter to first convert the payload and headers to a
+	 * {@code PubsubMessage} and then publish it.
+	 */
+	@Override
+	public <T> ListenableFuture<String> publish(String topic, T payload, Map<String, String> headers) {
+		return publish(topic, this.messageConverter.toPubSubMessage(payload, headers));
+	}
+
+	@Override
+	public <T> ListenableFuture<String> publish(String topic, T payload) {
+		return publish(topic, payload, null);
+	}
+
+	@Override
+	public ListenableFuture<String> publish(final String topic, PubsubMessage pubsubMessage) {
+		ApiFuture<String> publishFuture =
+				this.publisherFactory.createPublisher(topic).publish(pubsubMessage);
+
+		final SettableListenableFuture<String> settableFuture = new SettableListenableFuture<>();
+		ApiFutures.addCallback(publishFuture, new ApiFutureCallback<String>() {
+
+			@Override
+			public void onFailure(Throwable throwable) {
+				LOGGER.warn("Publishing to " + topic + " topic failed.", throwable);
+				settableFuture.setException(throwable);
+			}
+
+			@Override
+			public void onSuccess(String result) {
+				if (LOGGER.isDebugEnabled()) {
+					LOGGER.debug(
+							"Publishing to " + topic + " was successful. Message ID: " + result);
+				}
+				settableFuture.set(result);
+			}
+
+		});
+
+		return settableFuture;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+	}
+
+	public PublisherFactory getPublisherFactory() {
+		return this.publisherFactory;
+	}
+}
```

Comment:
Blank line above this line pls.

Reply:
Added blank line after last method declaration in every class.